drop-transfer/src/ws/client/mod.rs

```diff
@@ -44,43 +47,141 @@ pub enum ClientReq {
+                            let meta = fs::symlink_metadata(&fullpath).with_context(|| {
+                                format!("Failed to load file: {}", dbfile.file_id)
+                            })?;
+                            anyhow::ensure!(!meta.is_dir(), "Invalid file type");
+
+                            let file_id: FileId = dbfile.file_id.into();
+                            crate::File::new_to_send(
+                                dbfile.subpath.into(),
+                                fullpath,
+                                meta,
+                                file_id.clone(),
+                            )
+                            .with_context(|| format!("Failed to restore file {file_id} from DB"))
+                        })
+                        .collect::<Result<_, _>>()?;
+
+                    let xfer = crate::Transfer::new_with_uuid(
+                        transfer.peer.parse().context("Failed to parse peer IP")?,
+                        files,
+                        transfer.uuid,
+                        &state.config,
+                    )
+                    .context("Failed to create transfer")?;
+
+                    anyhow::Ok(xfer)
+                };
+
+                let xfer = match restore_transfer() {
+                    Ok(xfer) => xfer,
+                    Err(err) => {
+                        error!(logger, "Failed to restore transfer: {err}");
+                        return;
+                    }
+                };
+
+                run(state, xfer, logger).await
+            }
+        };
+
+        let logger = logger.clone();
+        let stop = stop.clone();
+
+        tokio::spawn(async move {
+            tokio::select! {
+                biased;
+
+                _ = stop.cancelled() => {
+                    warn!(logger, "Aborting transfer resume");
+                },
+                _ = task => (),
+            }
+        });
+    }
+}
+
+async fn connect_to_peer<const DB_INSERT: bool>(
+    state: &Arc<State>,
+    xfer: &crate::Transfer,
+    logger: &Logger,
+) -> ControlFlow<()> {
+    let (socket, ver) = match establish_ws_conn(state, xfer.peer(), logger).await {
         Ok(res) => res,
         Err(err) => {
             error!(logger, "Could not connect to peer {}: {}", xfer.id(), err);
 
             state
                 .event_tx
-                .send(Event::TransferFailed(xfer, err, false))
+                .send(Event::TransferFailed(xfer.clone(), err, false))
                 .await
                 .expect("Failed to send TransferFailed event");
 
-            return;
+            return ControlFlow::Break(());
         }
     };
 
     info!(logger, "Client connected, using version: {ver}");
 
+    if DB_INSERT {
+        if let Err(err) = state.storage.insert_transfer(&xfer.storage_info()).await {
+            error!(logger, "Failed to insert transfer into storage: {err}");
+        }
+    }
+
```

Comment:
There is a logic here. The connection failure is unrecoverable here.
But it's true that in case of failure, the transfer is not stored in DB which is probably not good?

Reply:
yeah, for one, we'd get a constraint check failed since the dispatch would react to the `TransferFailed` event, and possibly might be nice for the app teams to retrieve this failed information, worst case they could just ignore it

Reply:
yep, makes sense to insert the transfer always